File: app/src/androidTest/java/org/tasks/Freeze.java
package org.tasks;

import static org.tasks.time.DateTimeUtils.currentTimeMillis;

import org.tasks.time.DateTime;
import org.tasks.time.DateTimeUtils;

public class Freeze {

  public static Freeze freezeClock() {
    return freezeAt(currentTimeMillis());
  }

  public static Freeze freezeAt(DateTime dateTime) {
    return freezeAt(dateTime.getMillis());
  }

  public static Freeze freezeAt(long millis) {
    DateTimeUtils.setCurrentMillisFixed(millis);
    return new Freeze();
  }

  public static void thaw() {
    DateTimeUtils.setCurrentMillisSystem();
  }

  @SuppressWarnings("UnusedParameters")
  public void thawAfter(Snippet snippet) {
    thaw();
  }

  public void thawAfter(Runnable run) {
    try {
      run.run();
    } finally {
      thaw();
    }
  }
}

File: app/src/commonTest/java/org/tasks/Freeze.kt
package org.tasks

import org.tasks.time.DateTime
import org.tasks.time.DateTimeUtils

class Freeze {
    fun thawAfter(run: () -> Unit) {
        try {
            run()
        } finally {
            thaw()
        }
    }

    companion object {

        fun freezeClock(run: () -> Unit) {
            freezeAt(DateTimeUtils.currentTimeMillis()).thawAfter(run)
        }

        fun freezeAt(dateTime: DateTime, run: () -> Unit) {
            freezeAt(dateTime.millis, run)
        }

        fun freezeAt(timestamp: Long, run: () -> Unit) {
            freezeAt(timestamp).thawAfter(run)
        }

        fun freezeAt(dateTime: DateTime): Freeze {
            return freezeAt(dateTime.millis)
        }

        fun freezeAt(millis: Long): Freeze {
            DateTimeUtils.setCurrentMillisFixed(millis)
            return Freeze()
        }

        fun thaw() {
            DateTimeUtils.setCurrentMillisSystem()
        }
    }
}